### ISSUE TITLE
SFT inherits pre-batch-scaling learning rates from base checkpoint

### ISSUE THREAD

**Issue by Qwertyemma** [karpathy/nanochat#848]
I noticed a possible issue with how learning rates are passed from base training to SFT.

In base_train.py, the learning rates are scaled using batch_lr_scale before being passed to model.setup_optimizer():

unembedding_lr=args.unembedding_lr * batch_lr_scale,
embedding_lr=args.embedding_lr * batch_lr_scale,
scalar_lr=args.scalar_lr * batch_lr_scale,
matrix_lr=args.matrix_lr * batch_lr_scale,

However, user_config is created earlier from the original args:

user_config = vars(args).copy()

and the checkpoint later saves that user_config:

"user_config": user_config,

So the checkpoint contains the learning rates before batch_lr_scale is applied, while the optimizer during base training uses the scaled values.

In chat_sft.py, the SFT script reads the learning rates from the base checkpoint when they aren't explicitly provided:

pretrain_user_config = meta.get("user_config", {})

and then inherits values such as embedding_lr, unembedding_lr, and matrix_lr from it.

This means that when the base training batch size is different from B_REF, SFT can inherit a learning rate that is different from the effective learning rate used during base training.

I'm not completely sure whether user_config is intended to store the original user-provided values or the effective training configuration, so I wanted to raise this before making a PR.

If the intention is for SFT to inherit the effective base-training learning rates, I think the scaled values should be reflected in the checkpoint metadata, or alternatively the scaling information should be stored so that SFT can reconstruct the effective values.

**Comment by mira687** [comment]
Your reading is right, and the fix is smaller than either option you list — the scaling information is already in the checkpoint, three lines above where SFT reads the LRs.

`chat_sft.py` pulls `total_batch_size` from `meta` rather than from `pretrain_user_config`, and `base_train.py` saves that field *after* the auto-compute branch resolves it. So SFT can reconstruct the factor exactly, with no checkpoint format change:

```python
batch_lr_scale = (args.total_batch_size / 2**19) ** 0.5  # B_REF
```

The default in that same inheritance table is already `524288 == B_REF`, so a checkpoint without the field degrades to `1.0`.

On the intent question you raised: `user_config = vars(args).copy()  # for logging`, and it's saved as `"user_config": user_config, # inputs to the training script`. Both comments read as a record of what was typed, not as effective config — which points at SFT's inheritance being the bug rather than at user_config's contents.

One thing worth knowing before you size the PR: the magnitude isn't marginal. `total_batch_size` is clamped to a power of two, so `batch_ratio` is a power of two and `batch_lr_scale` moves in steps of √2. The first size step above B_REF leaves SFT inheriting LRs 1.41x low, not a few percent off.

Two smaller things while you're in there: `scalar_lr` is scaled in base_train but isn't in the SFT inheritance list at all, and weight decay doesn't need reconstructing — `chat_sft.py` hardcodes `weight_decay=0.0` on purpose, since pretraining ramps it to zero by the end.
